src/HealthGPS.Datastore/datamanager.cpp

```diff
@@ -1,33 +1,86 @@
+    constexpr const char *url_prefix = "http://example.com/";
+
+    const auto &uri_str = uri.string();
+    if (!uri_str.starts_with(url_prefix)) {
+        throw std::runtime_error(fmt::format("Unable to load URL: {}", uri_str));
+    }
+
+    // Strip URL prefix and load file from local filesystem
+    const auto schema_path = schema_directory / std::filesystem::path{uri.path().substr(1)};
+    auto ifs = std::ifstream{schema_path};
+    if (!ifs) {
+        throw std::runtime_error("Failed to read schema file");
+    }
+
+    return json::parse(ifs);
+}
+
+void validate_index(const std::filesystem::path &schema_directory, std::ifstream &ifs) {
+    // **YUCK**: We have to read in the data with jsoncons here rather than reusing
+    // the nlohmann-json representation :-(
+    const auto index = json::parse(ifs);
+
+    // Go back to start of file so caller can re-read it
+    ifs.seekg(0);
+
+    // Load schema
+    auto ifs_schema = std::ifstream{schema_directory / "data_index.json"};
+    if (!ifs_schema) {
+        throw std::runtime_error("Failed to load schema");
+    }
+
+    const auto resolver = [&schema_directory](const auto &uri) {
+        return resolve_uri(uri, schema_directory);
+    };
+    const auto schema = jsonschema::make_json_schema(json::parse(ifs_schema), resolver);
+
+    // Perform validation
+    schema.validate(index);
+}
+} // namespace
+
 namespace hgps::data {
 DataManager::DataManager(std::filesystem::path root_directory, VerboseMode verbosity)
     : root_{std::move(root_directory)}, verbosity_{verbosity} {
     auto full_filename = root_ / "index.json";
-    auto ifs = std::ifstream{full_filename, std::ifstream::in};
-    if (ifs) {
-        index_ = nlohmann::json::parse(ifs);
-        if (!index_.contains("version")) {
-            throw std::runtime_error("File-based store, invalid definition missing schema version");
-        }
-
-        auto version = index_["version"].get<int>();
-        if (version != 2) {
-            throw std::runtime_error(fmt::format(
-                "File-based store, index schema version: {} mismatch, supported: 2", version));
-        }
-    } else {
+    auto ifs = std::ifstream{full_filename};
+    if (!ifs) {
         throw std::invalid_argument(
             fmt::format("File-based store, index file: '{}' not found.", full_filename.string()));
     }
+
+    // Validate against schema
+    const auto schema_directory = get_program_directory() / "schemas" / "v1";
+    validate_index(schema_directory, ifs);
+
+    index_ = nlohmann::json::parse(ifs);
+
+    if (!index_.contains("version")) {
+        throw std::runtime_error("File-based store, invalid definition missing schema version");
+    }
+
+    auto version = index_["version"].get<int>();
+    if (version != 2) {
+        throw std::runtime_error(fmt::format(
+            "File-based store, index schema version: {} mismatch, supported: 2", version));
+    }
```

Comment:
Indeed they are not! Unfortunately though `jsoncons` doesn't always give particularly helpful error messages, so maybe we don't want to rip them out just yet. We could move the checking with the schema till after we've run these checks, so that it's basically used as a fallback... What do you think?